### PR TITLE
[SPARK-21223] Change fileToAppInfo in FsHistoryProvider to fix concurrent issue.

### PR DESCRIPTION
**Pull request by zenglinxi0615**
# What issue does this PR address ?
Jira:https://issues.apache.org/jira/browse/SPARK-21223
fix the Thread-safety issue in FsHistoryProvider
Currently, Spark HistoryServer use a HashMap named fileToAppInfo in class FsHistoryProvider to store the map of eventlog path and attemptInfo. 
When use ThreadPool to Replay the log files in the list and merge the list of old applications with new ones, multi thread may update fileToAppInfo at the same time, which may cause Thread-safety issues, such as  falling into an infinite loop because of calling resize func of the hashtable.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 1]
```diff
@@ -122,7 +123,7 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
   @volatile private var applications: mutable.LinkedHashMap[String, FsApplicationHistoryInfo]
     = new mutable.LinkedHashMap()
 
-  val fileToAppInfo = new mutable.HashMap[Path, FsApplicationAttemptInfo]()
+  val fileToAppInfo = new java.util.concurrent.ConcurrentHashMap[Path, FsApplicationAttemptInfo]()
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Import this class

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 2]
```diff
@@ -321,7 +322,7 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       // scan for modified applications, replay and merge them
       val logInfos: Seq[FileStatus] = statusList
         .filter { entry =>
-          val prevFileSize = fileToAppInfo.get(entry.getPath()).map{_.fileSize}.getOrElse(0L)
+          val prevFileSize = fileToAppInfo.asScala.get(entry.getPath()).map{_.fileSize}.getOrElse(0L)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: `asScala` produces a wrapper, right? in which case this can be called once, at declaration above. If it produces a clone, this doesn't work anyway!
But why is this needed anyway? you can `get` elements from the Java map already.

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala (hunk 2 above)
Comment (zenglinxi0615) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: @srowen thanks for your suggestions, I have made some modifications, could you take a look when you have time?

**Comment by jerryshao** [comment]
Can you please fix your PR description to remove template words? BTW did you see the inconsistent issue here?

**Comment by zenglinxi0615** [comment]
@jerryshao do you mean that after fileToAppInfo.get(entry.getPath()) return a value, other threads may add or change the value of entry.getPath(), which cause an inconsistent issue?

**Comment by jerryshao** [comment]
You're saying you're fixing a threading issue, so my question is did you see any inconsistent behavior regarding to threading contention?

**Comment by zenglinxi0615** [comment]
@jerryshao  actually, this threading issue cause an infinite loop when we restart historyserver and replaying event logs of spark apps. you can see the jstack log in attachments of SPARK-21223.

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 3]
```diff
@@ -321,7 +321,11 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
       // scan for modified applications, replay and merge them
       val logInfos: Seq[FileStatus] = statusList
         .filter { entry =>
-          val prevFileSize = fileToAppInfo.get(entry.getPath()).map{_.fileSize}.getOrElse(0L)
+          var prevFileSize = 0L
+          val fileInfo = fileToAppInfo.get(entry.getPath())
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: Simpler:
```
val fileInfo = fileToAppInfo.get(entry.getPath())
val prevFileSize = if (fileInfo != null) fileInfo.fileSize else 0L
```

**Review by srowen — CHANGES_REQUESTED** [review]

**Comment by zenglinxi0615** [comment]
@srowen thanks for your suggestions again! and should I address the problem of SPARK-13988 in this PR?

**Comment by jerryshao** [comment]
What's the issue of SPARK-13988?

**Comment by zenglinxi0615** [comment]
sorry, it's a typing error, i mean the related JIRA: SPARK-21078.

**Comment by jerryshao** [comment]
let's do this in another PR, seems a different threading issue.

Also would you please change the PR title like: Change fileToAppInfo in FsHistoryProvider to fix concurrent issue.

**Review by jiangxb1987 — APPROVED** [review]
LGTM, cc @cloud-fan

**Comment by jiangxb1987** [comment]
Also Please change the title as mentioned above. Thanks!

**Comment by cloud-fan** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 4]
```diff
@@ -19,7 +19,7 @@ package org.apache.spark.deploy.history
 
 import java.io.{FileNotFoundException, IOException, OutputStream}
 import java.util.UUID
-import java.util.concurrent.{Executors, ExecutorService, Future, TimeUnit}
+import java.util.concurrent.{Executors, ExecutorService, Future, TimeUnit, ConcurrentHashMap}
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: import java.util.concurrent.{ConcurrentHashMap, Executors, ExecutorService, Future, TimeUnit}

**Comment by zenglinxi0615** [comment]
test please

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master
